node/src/components/network.rs

```diff
@@ -0,0 +1,610 @@
+        //        can't re-enter this block unless we implement re-bootstrapping for example.
+        if self.is_isolated() {
+            if self.is_bootstrap_node() {
+                info!(
+                    "{}: failed to bootstrap to any other nodes, but continuing to run as we are a \
+                    bootstrap node", self.our_id
+                );
+            } else {
+                // Note that we could retry the connection to other nodes, but for now we just
+                // leave it up to the node operator to restart.
+                return fatal!(
+                    effect_builder,
+                    "{}: failed to connect to any known node, now isolated",
+                    self.our_id
+                );
+            }
+        }
+        Effects::new()
+    }
+
+    fn handle_new_listen_address(&mut self, address: Multiaddr) -> Effects<Event<P>> {
+        self.listening_addresses.push(address);
+        info!(
+            "{}: listening on {}",
+            self.our_id,
+            DisplayIter::new(self.listening_addresses.iter())
+        );
+        Effects::new()
+    }
+
+    fn handle_expired_listen_address(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        address: Multiaddr,
+    ) -> Effects<Event<P>> {
+        self.listening_addresses.retain(|addr| *addr != address);
+        if self.listening_addresses.is_empty() {
+            return fatal!(effect_builder, "no remaining listening addresses");
+        }
+        debug!(%address, "{}: listening address expired", self.our_id);
+        Effects::new()
+    }
+
+    fn handle_listener_closed(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        _addresses: Vec<Multiaddr>,
+        reason: Result<(), io::Error>,
+    ) -> Effects<Event<P>> {
+        // If the listener closed without an error, we're already shutting down the server.
+        // Otherwise, we need to kill the node as it cannot function without a listener.
+        match reason {
+            Err(error) => fatal!(effect_builder, "listener closed: {}", error),
+            Ok(()) => {
+                debug!("{}: listener closed", self.our_id);
+                Effects::new()
+            }
+        }
+    }
+
+    fn handle_listener_error(&self, error: io::Error) -> Effects<Event<P>> {
+        debug!(%error, "{}: non-fatal listener error", self.our_id);
+        Effects::new()
+    }
+
+    /// Returns whether or not this node has been isolated.
+    ///
+    /// An isolated node has no chance of recovering a connection to the network and is not
+    /// connected to any peer.
+    fn is_isolated(&self) -> bool {
+        self.known_addresses
+            .values()
+            .all(|state| *state == ConnectionState::Failed)
+    }
+
+    /// Returns whether or not this node is listed as a bootstrap node.
+    fn is_bootstrap_node(&self) -> bool {
```

Comment:
Can we get rid of the bootstrapping logic altogether (the distinction between non-bootstrap and bootstrap nodes)? Ideally, it should not make a difference except for the contents of the initial address pool.

Reply:
Well, just now, we panic if we become "isolated", which is probably the right thing to do.  However, if we're a bootstrap node, or at least the very first node on the network, then we have to _not_ panic under such circumstances.

Reply:
Fair enough.